### PR TITLE
kernel: version: option to enable GNU build ID

### PR DESCRIPTION
**Pull request by JordanYates**
Adds a Kconfig option to enable the generation of the GNU build ID from the final `zephyr.elf` file. This build ID can be used to uniquely identify a firmware image.

The `--build-id` option is available for the `ld` and `ldd` linkers, which are used by all in-tree compilers except `arcmwdt` and `armclang`.

Inspired by the following blog post: https://interrupt.memfault.com/blog/gnu-build-id-for-firmware

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

### CODE REVIEW

**Comment by Laczen** [comment]
@JordanYates, would it be possible to make the build-id available outside of the image? E.g. could it be read during/before firmware upgrade?

**Comment by JordanYates** [comment]
> @JordanYates, would it be possible to make the build-id available outside of the image? E.g. could it be read during/before firmware upgrade?

This would require either the section is placed at a constant memory address (which is probably unworkable) or being part of the image metadata of whatever upgrade mechanism is being used.

From a quick look MCUBoot image trailers would probably be suitable for this purpose. This would require the signing tool to be updated to add the trailer (And defining the TLV type). I imagine MCUBoot would make the argument that they already have `major.minor.path` versioning though, so why have another version number.

**Comment by Laczen** [comment]
> > @JordanYates, would it be possible to make the build-id available outside of the image? E.g. could it be read during/before firmware upgrade?
> 
> This would require either the section is placed at a constant memory address (which is probably unworkable) or being part of the image metadata of whatever upgrade mechanism is being used.

It could be added to the front of the image reclaming part of the 512 byte header that is added to support e.g. mcuboot.

> 
> From a quick look MCUBoot image trailers would probably be suitable for this purpose. This would require the signing tool to be updated to add the trailer (And defining the TLV type). I imagine MCUBoot would make the argument that they already have `major.minor.path` versioning though, so why have another version number.

Adding it to the header could remove any need of TLV processing for the image to access the information. A solution could be to add it just after the 32 byte header that MCUBoot uses (if mcuboot is used). One could think of creating a zephyr header that leaves space for a MCUBoot header if required and after that placing extra zephyr image information.

**Comment by nordicjm** [comment]
> Adding it to the header could remove any need of TLV processing for the image to access the information

I was under the impression that the header was the same TLV system as the footer, the difference being about encryption and if it's included in the final signature or not?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Why is this needed ?

We already create `version.h` which looks like this:
```
cat build/zephyr/include/generated/version.h
#ifndef _KERNEL_VERSION_H_
#define _KERNEL_VERSION_H_

/* KERNEL and ZEPHYR_VERSION  values come from cmake/version.cmake
 * BUILD_VERSION  will be 'git describe', alternatively user defined BUILD_VERSION.
 */

#define ZEPHYR_VERSION_CODE 197219
#define ZEPHYR_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define KERNELVERSION          0x3026300
#define KERNEL_VERSION_NUMBER  0x30263
#define KERNEL_VERSION_MAJOR   3
#define KERNEL_VERSION_MINOR   2
#define KERNEL_PATCHLEVEL      99
#define KERNEL_VERSION_STRING  "3.2.99"

#define BUILD_VERSION          zephyr-v3.2.0-1066-g217528f2dea8

#endif /* _KERNEL_VERSION_H_ */
```

which uses the `VERSION` file and `git describe`:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/version.cmake
https://github.com/zephyrproject-rtos/zephyr/blob/217528f2dea83a90f5f4114f79177fa9d2c20e5f/cmake/gen_version_h.cmake#L9
but also provide users with the possibility of specifying a custom `VERSION`.
https://github.com/zephyrproject-rtos/zephyr/blob/217528f2dea83a90f5f4114f79177fa9d2c20e5f/CMakeLists.txt#L474-L475

All of which is toolchain independent.

Why is a GNU specific feature needed, and even one that is `default y`.
What are the benefit of one more way of achieving the same ?

**Comment by JordanYates** [comment]
> What are the benefit of one more way of achieving the same ?

Mostly because none of the methods you provide can guarantee that an image you are communicating with is the image you actually think it is. 

`git describe` is dependent on the commit ID changing, you can build an infinite number of variations of hello_world and git will happily tell you they are all the same.

Same deal with `-DBUILD_VERSION`. Maybe you're under pressure to release something and forget to change the define, or you fat finger it, or any number of other potential reasons.

The GNU build ID tells you, without any doubt, the binary running on this device is the same as the binary I am inspecting on my computer.

**Comment by nordicjm** [comment]
Adding @stephanosio for comment. I don't think we want to introduce anything that is limited to a GCC-based compiler only, that is far too limiting.

**Comment by JordanYates** [comment]
> Adding @stephanosio for comment. I don't think we want to introduce anything that is limited to a GCC-based compiler only, that is far too limiting.

The limitation is on the linker, not the compiler. The option is supported by `ld` and `lld`, which is used by every in-tree compiler except `arcmwdt` and `armclang`.

**Comment by JordanYates** [comment]
>>The limitation is on the linker, not the compiler. The option is supported by ld and lld, which are used by every in-tree compiler except arcmwdt and armclang.

> And IAR?

I don't see it under `cmake/compiler` or in the Zephyr SDK, so I'm not sure its fair to call that in-tree. Honestly I have never used it so I have no idea what options it may or may not support.

**Review by stephanosio — COMMENTED** [review]
> The GNU build ID tells you, without any doubt, the binary running on this device is the same as the binary I am inspecting on my computer.

I can see that is a very useful feature, but I am not convinced GNU build ID is the right solution for it.

As you and others pointed out, this feature only works with GNU and LLVM toolchains and will not work with any other toolchains in its current form.

IMHO, we should try to provide a more generic solution to this problem. For example, a Python script that injects the hash of the pre-stage ELF file to the final ELF file (basically the same thing that the ld does for GNU build ID).

**Comment by JordanYates** [comment]
> IMHO, we should try to provide a more generic solution to this problem. For example, a Python script that injects the hash of the pre-stage ELF file to the final ELF file (basically the same thing that the ld does for GNU build ID).

I suspect this is more complicated than just calculating a hash of the `.elf` file, and also raises the question of which link stage are you calculating the hash over? Can you guarantee that having a hash of one of the "pre" stages is sufficient? Or do we add yet-another-link-stage?

My personal opinion is that `ld` and `lld` covers a sufficiently large segment of the userbase that adding the option, which only requires modifying a linker flag, provides sufficient value until a user of the non-supported toolchains is motivated to write the (probably complex) python parser/injecter. Or asks their friendly neighbourhood compiler vendor to implement a `--build-id` flag.

**Comment by ceolin** [comment]
I liked it that is a good feature, and as @JordanYates mentioned, the current versioning does not give a whole picture of what was built, as Zephyr (almost by lucky) has reproducible builds, this feature becomes really powerful. I'm in favor in the proposed implementation, it is simple enough and probably covers most builds. To make it generic we problably would need to reserve an area and add one additional final step in the build that overrides it with the unique ID.

For me this feature is just simple and good enough, if one needs something more robust probably will go to a solution that involves sign the image.

**Comment by JordanYates** [comment]
Looking to close this out with either accepted or rejected.
@tejlmand do you have any comments on my response to your question?

**Comment by nordicjm** [comment]
I don't think should go in being limited to GCC only, does the xtensa compiler support it? ARC? For most devices we have the GCC zephyr toolchain but there was (not sure if there still is) talk about supporting other compilers and this just introduces too much of headache in my mind, and I don't see this being *essential* as per the requirements listed on https://github.com/zephyrproject-rtos/zephyr/issues/52226

File: Kconfig.zephyr [hunk 1]
```diff
@@ -200,6 +200,18 @@ config KERNEL_ENTRY
 	help
 	  Code entry symbol, to be set at linking phase.
 
+config TOOLCHAIN_SUPPORTS_GNU_BUILD_ID
+	bool
+	default y
```
Comment (attie-argentum) [inline Kconfig.zephyr:205]: I'd make this `default n`, and explicitly set it `y` for those that support it.

**Review by attie-argentum — CHANGES_REQUESTED** [review]
Following on from #54522 and comments above... I'd suggest that this is suitably "robust, low effort and inclusive" to be worth considering seriously - but I fully take the point regarding "GNUisms".

In my opinion, the alternative to this small patch would be a larger and potentially fragile solution, which may still not suit 100% of users.

File: kernel/version.c [hunk 2]
```diff
@@ -19,3 +20,24 @@ uint32_t sys_kernel_version_get(void)
...
+	uint32_t name_sz;
+	uint32_t desc_sz;
+	uint32_t type;
+	uint8_t data[];
+};
+
+/* Variable instantiated by the linker */
+extern const struct elf_note_section __g_note_build_id;
+
+const uint8_t *sys_gnu_build_id_get(void)
```
Comment (attie-argentum) [inline kernel/version.c:37]: In the event that this expands out into supporting other toolchains, the build ID may be differently sized... I'd suggest an interface more like this.

```suggestion
void sys_build_id_get(const uint8_t **buf, size_t *len)
```

**Comment by Laczen** [comment]
@attie-argentum, have you seen the recent proposal from @yonsch (#54464) ?

**Comment by attie-argentum** [comment]
@Laczen - I had not, definitely worth tying these all together, thanks!

**Comment by yonsch** [comment]
For anyone still interested, another approach to implement build ID: #63350

**Pull request closed without merging**